server/market/order/order.go

```diff
@@ -0,0 +1,326 @@
+// Copyright (c) 2019, The Decred developers
+// See LICENSE for details.
+
+// Package order defines the Order and Match types used throughout the DEX.
+package order
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"math"
+	"time"
+
+	"github.com/decred/dcrd/crypto/blake256"
+	"github.com/decred/dcrdex/server/account"
+)
+
+// OrderIDSize defines the length in bytes of an OrderID.
+const OrderIDSize = blake256.Size
+
+// OrderID is the unique identifier for each order.
+type OrderID [OrderIDSize]byte
+
+// String returns a hexadecimal representation of the OrderID. String implements
+// fmt.Stringer.
+func (oid OrderID) String() string {
+	return hex.EncodeToString(oid[:])
+}
+
+// OrderType distinguishes the different kinds of orders (e.g. limit, market,
+// cancel).
+type OrderType uint8
+
+// The different OrderType values.
+const (
+	LimitOrderType OrderType = iota
+	MarketOrderType
+	CancelOrderType
+)
+
+// TimeInForce indicates how limit order execution is to be handled. That is,
+// when the order is not immediately matched during processing of the order's
+// epoch, the order may become a standing order or be revoked without a fill.
+type TimeInForce uint8
+
+// The TimeInForce is either ImmediateTiF, which prevents the order from
+// becoming a standing order if there is no match during epoch processing, or
+// StandingTiF, which allows limit orders to enter the order book if not
+// immediately matched during epoch processing.
+const (
+	ImmediateTiF TimeInForce = iota
+	StandingTiF
+)
+
+// Order specifies the methods required for a type to function as a DEX order.
+// See the concrete implementations of MarketOrder, LimitOrder, and CancelOrder.
+type Order interface {
+	// ID computes the Order's ID from its serialization as per the spec.
+	ID() OrderID
+
+	// UID gives the string representation of the order ID. It is named to
+	// reflect the intent of providing a unique identifier.
+	UID() string
+
+	// Serialize marshals the order as per the spec.
+	Serialize() []byte
+
+	// SerializeSize gives the length of the serialized order in bytes.
+	SerializeSize() int
+
+	// Type indicates the Order's type (e.g. LimitOrder, MarketOrder, etc.).
+	Type() OrderType
+
+	// Time returns the Order's server time, when it was received by the server.
+	Time() int64
+
+	// Remaining computes the unfilled amount of the order.
+	Remaining() uint64
+}
+
+// An order's ID is computed as the Blake-256 hash of the serialized order.
+func calcOrderID(order Order) OrderID {
+	return blake256.Sum256(order.Serialize())
+}
+
+// UTXO is the interface required to be satisfied by any asset's implementation
+// of a UTXO type.
+type UTXO interface {
+	TxHash() []byte
+	Vout() uint32
+	Serialize() []byte
+	SerializeSize() int
```

Comment:
I think I misunderstood your previous comment about this.  This order package must have either a way to obtain a UTXO's serialization or have it's own way of doing the serialization from the TxHash and Vout.  But I took your previous comment about this issue to me you felt that serialization was a higher level (DEX) responsibility, meaning the order package should leave that up to whoever provides the UTXO by providing a Serialize() method.

On the other hand, I don't have a problem letting the order package perform the UTXO serialization from the hash and vout as per the spec.

Reply:
I imagine the `UTXO` interface to have `TxHash` and `Vout` methods. The higher level modules (such as the order types' serialize methods) would take care of putting them together. If we need a `HashSize` method somewhere, we can do that, but I'm not sure we would need it. The hash will likely have to be retrieved through the backend during initial order validation anyway.

My motivation is to keep these DEX-level specifics out of the asset backends so that if these details ever change, there are no changes required to the backends. As an example more specific to this discussion, if some future modification of the DEX decided that it's important that the vout go first in the serialization rather than the tx hash, we don't want to require all of the backends to require updating. A DEX-level detail like that should not impact asset backend implementation, whenever possible. 

I don't foresee us ever needing that particular change. More just trying to establish the boundaries of concern so that things stay modular.

Reply:
Understood, thanks for clarifying how you envisioned that.  `TxHash() []byte` gives what's needed, I just misunderstood your comment on the UTXO issue on your review from several days ago.

Reply:
Resolved in https://github.com/decred/dcrdex/pull/15/commits/fbf0b5cdfaeb23cd04e7812fa3f229be1021dc33.